mqtt/mqtt-edgehub/src/command.rs

```diff
@@ -0,0 +1,249 @@
+    address: String,
+}
+
+impl IoSource for BrokerConnection {
+    type Io = TcpStream;
+    type Error = std::io::Error;
+    type Future = BoxFuture<'static, Result<(TcpStream, Option<String>), std::io::Error>>;
+
+    fn connect(&mut self) -> Self::Future {
+        let address = self.address.clone();
+        Box::pin(async move {
+            let io = TcpStream::connect(address).await;
+            io.map(|io| (io, None))
+        })
+    }
+}
+
+pub struct CommandHandler {
+    broker_handle: BrokerHandle,
+    client: Client<BrokerConnection>,
+}
+
+impl CommandHandler {
+    pub async fn new(
+        broker_handle: BrokerHandle,
+        address: String,
+        device_id: String,
+    ) -> Result<Self, InitializationError> {
+        let client_id = format!("{}/$edgeHub/$broker/$control", device_id);
+
+        let broker_connection = BrokerConnection { address };
+
+        let mut client = Client::new(
+            Some(client_id),
+            None,
+            None,
+            broker_connection,
+            Duration::from_secs(1),
+            Duration::from_secs(60),
+        );
+
+        let qos = proto::QoS::AtLeastOnce;
+        client
+            .subscribe(proto::SubscribeTo {
+                topic_filter: TOPIC_FILTER.to_string(),
+                qos,
+            })
+            .map_err(InitializationError::SubscribeFailure)?;
+
+        wait_for_suback(&mut client).await?;
+
+        Ok(CommandHandler {
+            broker_handle,
+            client,
+        })
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn run(mut self) {
+        while let Some(Ok(event)) = self.client.next().await {
+            if let Err(e) = self.handle_event(event).await {
+                warn!(message = "failed to disconnect client", error = %e);
+            }
+        }
+    }
+
+    async fn handle_event(&mut self, event: Event) -> Result<(), DisconnectClientError> {
+        if let Event::Publication(publication) = event {
+            let topic_name = publication.topic_name.as_ref();
+            let client_id = parse_client_id(topic_name)?;
+
+            info!("received disconnection request for client {}", client_id);
+
+            if let Err(e) = self
+                .broker_handle
+                .send(Message::System(SystemEvent::ForceClientDisconnect(
+                    client_id.into(),
+                )))
+                .await
+            {
+                return Err(DisconnectClientError::SignalError(e));
+            }
+
+            info!("succeeded sending broker signal to disconnect client");
+        }
+
+        Ok(())
+    }
+}
+
+async fn wait_for_suback(client: &mut Client<BrokerConnection>) -> Result<(), InitializationError> {
```

Comment:
I want this to throw an initialization error if this suback wait "times out" of some arbitrary threshold. Any suggestions for how to do it? 

```while client.next()``` body waits until the next event is received so you can't break out of the loop. I was thinking of spawning a new tokio thread and then shutting it down when it timed out. Is this a good idea?

Reply:
talked with Denis and we decided that it is fine if the suback wait doesn't have a timeout, since the client will shutdown if the broker fails.